utils/test_results.py

```diff
@@ -64,38 +73,134 @@ def aggregate_test_results(
+                + Cast(Sum(F("fail_count")), output_field=FloatField())
+            )
+        ),
+        flake_rate=(
+            Cast(Sum(F("flaky_fail_count")), output_field=FloatField())
+            / (
+                Cast(
+                    Sum(F("pass_count")),
+                    output_field=FloatField(),
+                )
+                + Cast(Sum(F("fail_count")), output_field=FloatField())
+            )
+        ),
+        updated_at=Max("latest_run"),
+        commits_where_fail=ArrayLength(Array(Subquery(commits_where_fail_sq))),
+        last_duration=Subquery(latest_duration_sq),
+        avg_duration=Avg("avg_duration_seconds"),
+        name=F("test__name"),
+    )
+
+    return aggregation_of_test_results
+
+
+def generate_test_results_aggregates(
+    repoid: int, history: dt.timedelta | None = None
+) -> dict[str, float | int] | None:
+    repo = Repository.objects.get(repoid=repoid)
+    time_ago = (
+        (dt.datetime.now(dt.UTC) - history) if history is not None else thirty_days_ago
+    )
+
+    totals = DailyTestRollup.objects.filter(
+        repoid=repoid, date__gt=time_ago, branch=repo.branch
+    )
+
+    num_tests = totals.distinct("test_id").count()
+
+    slow_test_threshold = floor(num_tests * (100 - SLOW_TEST_PERCENTILE) * 0.01)
+
+    if slow_test_threshold == 0:
+        if num_tests == 1:
+            slow_test_threshold = 1
+        else:
+            return None
+
+    slowest_test_ids = (
+        totals.values("test")
+        .annotate(
+            runtime=Avg("avg_duration_seconds")
+            * (Sum("pass_count") + Sum("fail_count"))
         )
-        .values(
-            "failure_rate",
-            "commits_where_fail",
-            "avg_duration",
-            "name",
-            "updated_at",
+        .order_by("-runtime")
+        .values("test_id")[0:slow_test_threshold]
+    )
+
+    slowest_tests_duration = (
+        totals.filter(test_id__in=slowest_test_ids)
+        .values("repoid")
+        .annotate(
+            slowest_tests=Avg("avg_duration_seconds")
+            * (Sum("pass_count") + Sum("fail_count"))
         )
+        .values("slowest_tests")
+    )
+
+    test_headers = totals.values("repoid").annotate(
+        total_run_time=Avg("avg_duration_seconds")
+        * (Sum("pass_count") + Sum("fail_count")),
+        slowest_tests_duration=slowest_tests_duration,
+        skips=Sum("skip_count"),
+        fails=Sum("fail_count"),
     )
 
-    return failure_rates_queryset
+    if test_headers:
+        return test_headers[0]
+    else:
+        return None
+
+
+def generate_flake_aggregates(repoid: int, history: dt.timedelta | None = None):
+    repo = Repository.objects.get(repoid=repoid)
+    time_ago = (dt.date.today() - history) if history is not None else thirty_days_ago
+
+    flakes = Flake.objects.filter(
+        Q(repository_id=repoid)
+        & (Q(end_date__date__isnull=True) | Q(end_date__date__gt=time_ago))
```

Comment:
can you remind me again when end_date would be null?

Reply:
end_date is null when the flake is still active